src/BizHawk.Emulation.Common/Base Implementations/ControllerDefinition.cs

```diff
@@ -21,38 +19,48 @@ public class ControllerDefinition
 
 		private bool _mutable = true;
 
-		private IReadOnlyList<IReadOnlyList<string>> _orderedControls = null;
-
-		private IReadOnlyList<string> _orderedControlsFlat = null;
+		private IReadOnlyList<IReadOnlyList<(string, AxisSpec?)>>? _orderedControls;
 
 		/// <summary>starts with console buttons, then each player's buttons individually</summary>
-		public IReadOnlyList<IReadOnlyList<string>> ControlsOrdered
+		public IReadOnlyList<IReadOnlyList<(string Name, AxisSpec? AxisSpec)>> ControlsOrdered
 		{
 			get
 			{
 				if (_orderedControls is not null) return _orderedControls;
 				if (!_mutable) return _orderedControls = GenOrderedControls();
-				const string ERR_MSG = "this " + nameof(ControllerDefinition) + " has not yet been built and sealed, so it is not safe to enumerate this while it could still be mutated";
+				const string ERR_MSG = $"this {nameof(ControllerDefinition)} has not yet been built and sealed, so it is not safe to enumerate this while it could still be mutated";
 				throw new InvalidOperationException(ERR_MSG);
 			}
 		}
 
 		public readonly string Name;
 
-		public IReadOnlyList<string> OrderedControlsFlat => _orderedControlsFlat ??= ControlsOrdered.SelectMany(static s => s).ToList();
+		private Dictionary<string, char>? _mnemonicsCache;
+		public IReadOnlyDictionary<string, char>? MnemonicsCache => _mnemonicsCache;
+
+		/// <remarks>
+		/// TODO: this should probably be called in <see cref="MakeImmutable"/>,
+		/// but the needed Bk2MnemonicsLookup is in Client.Common
+		/// </remarks>

```

Comment:
???
Move the file?

Reply:
iirc I didn't because of semantics: bk2 is a movie thing and I wasn't sure if it was proper for the file to be in BizHawk.Emulation.Common instead. But looking at your commit it makes sense to move it.